File: ChatP2P.Core/LocalDb.Extensions.vb
Option Strict On
Imports System.Data
Imports System.Globalization

Namespace ChatP2P.Core

    ' Helpers DB additionnels (alignés roadmap) :
    ' - KV (Set/Get)
    ' - PeerKeys (Add / Get liste pour UI Révocations)
    ' - Purge messages plus anciens que N jours
    ' - Lecture simple des N derniers messages
    ' S’appuie sur les helpers de LocalDb.vb : ExecNonQuery / ExecScalar / Query / P

    Public Module LocalDbExtensions

        ' --- KV ---
        Public Sub KvSet(k As String, v As String)
            If String.IsNullOrWhiteSpace(k) Then Exit Sub
            LocalDb.ExecNonQuery(
                "INSERT INTO Kv(K,V) VALUES(@k,@v) ON CONFLICT(K) DO UPDATE SET V=excluded.V;",
                LocalDb.P("@k", k), LocalDb.P("@v", If(v, ""))
            )
        End Sub

        Public Function KvGet(k As String) As String
            If String.IsNullOrWhiteSpace(k) Then Return Nothing
            Dim o = LocalDb.ExecScalar(Of Object)("SELECT V FROM Kv WHERE K=@k;", LocalDb.P("@k", k))
            If o Is Nothing OrElse o Is DBNull.Value Then Return Nothing
            Return Convert.ToString(o, CultureInfo.InvariantCulture)
        End Function

        ' --- PeerKeys ---
        Public Sub PeerKeyAdd(peerName As String, kind As String, publicKey As Byte(), Optional note As String = Nothing)
            If String.IsNullOrWhiteSpace(peerName) OrElse String.IsNullOrWhiteSpace(kind) OrElse publicKey Is Nothing Then Exit Sub
            Dim nowIso = DateTime.UtcNow.ToString("o", CultureInfo.InvariantCulture)
            LocalDb.ExecNonQuery(
                "INSERT INTO PeerKeys(PeerName, Kind, Public, CreatedUtc, Revoked, RevokedUtc, Note)
                 VALUES(@p,@k,@pub,@ts,0,NULL,@n);",
                LocalDb.P("@p", peerName), LocalDb.P("@k", kind),
                LocalDb.P("@pub", publicKey), LocalDb.P("@ts", nowIso),
                LocalDb.P("@n", If(note, ""))
            )
        End Sub

        Public Function PeerKeysGet(peerName As String) As DataTable
            If String.IsNullOrWhiteSpace(peerName) Then Return New DataTable()
            Return LocalDb.Query(
                "SELECT Id, Kind, hex(Public) AS PubHex, CreatedUtc, Revoked, RevokedUtc, Note
                 FROM PeerKeys WHERE PeerName=@p ORDER BY Id DESC;",
                LocalDb.P("@p", peerName)
            )
        End Function

        ' --- Purge Messages ---
        Public Sub MessagesPurgeOlderThan(days As Integer)
            If days <= 0 Then Exit Sub
            Dim threshold = DateTime.UtcNow.AddDays(-days).ToString("o", CultureInfo.InvariantCulture)
            LocalDb.ExecNonQuery("DELETE FROM Messages WHERE CreatedUtc < @t;", LocalDb.P("@t", threshold))
        End Sub

        ' --- Historique simple (N derniers) ---
        Public Function MessagesGetLast(peer As String, take As Integer) As DataTable
            If String.IsNullOrWhiteSpace(peer) OrElse take <= 0 Then Return New DataTable()
            Return LocalDb.Query(
                "SELECT Sender, Body, IsP2P, Direction, CreatedUtc
                 FROM Messages WHERE PeerName=@p
                 ORDER BY Id DESC LIMIT @n;",
                LocalDb.P("@p", peer), LocalDb.P("@n", take)
            )
        End Function

        ' ====== Ed25519 Identity & Peer pubkeys (TOFU) ======

        Public Sub IdentityEnsureEd25519(ByRef pk As Byte(), ByRef sk As Byte())
            Dim dt = LocalDb.Query("SELECT Ed25519Pub, Ed25519Priv FROM Identities WHERE Id=1;")
            If dt.Rows.Count > 0 AndAlso Not IsDBNull(dt.Rows(0)(0)) AndAlso Not IsDBNull(dt.Rows(0)(1)) Then
                pk = CType(dt.Rows(0)(0), Byte())
                sk = CType(dt.Rows(0)(1), Byte())
                Exit Sub
            End If
            Dim gen = ChatP2P.Crypto.Ed25519Util.GenerateKeyPair()
            pk = gen.pk : sk = gen.sk
            LocalDb.ExecNonQuery(
        "UPDATE Identities SET Ed25519Pub=@p, Ed25519Priv=@s WHERE Id=1;",
        LocalDb.P("@p", pk), LocalDb.P("@s", sk))
        End Sub

        Public Function PeerGetEd25519(pubName As String) As Byte()
            Dim dt = LocalDb.Query("SELECT Public FROM PeerKeys WHERE PeerName=@p AND Kind='Ed25519' AND Revoked=0 ORDER BY Id DESC LIMIT 1;",
                           LocalDb.P("@p", pubName))
            If dt.Rows.Count = 0 OrElse IsDBNull(dt.Rows(0)(0)) Then Return Nothing
            Return CType(dt.Rows(0)(0), Byte())
        End Function

        Public Sub PeerSetEd25519_Tofu(pubName As String, pk As Byte())
            If pk Is Nothing Then Exit Sub
            LocalDb.EnsurePeer(pubName)
            Dim nowIso = DateTime.UtcNow.ToString("o", Globalization.CultureInfo.InvariantCulture)
            LocalDb.ExecNonQuery(
        "INSERT INTO PeerKeys(PeerName,Kind,Public,CreatedUtc,Revoked) VALUES(@p,'Ed25519',@k,@ts,0);",
        LocalDb.P("@p", pubName), LocalDb.P("@k", pk), LocalDb.P("@ts", nowIso))
        End Sub

        Public Sub PeerMarkVerified(name As String)
            LocalDb.ExecNonQuery("UPDATE Peers SET Verified=1 WHERE Name=@n;", LocalDb.P("@n", name))
        End Sub




    End Module

End Namespace

File: ChatP2P.Server/LocalDbExtensions.cs
using System;
using System.Data;
using System.Data.SQLite;

namespace ChatP2P.Server
{
    /// <summary>
    /// Port C# of VB.NET LocalDbExtensions - Extended database operations
    /// </summary>
    public static class LocalDbExtensions
    {
        public static void KvSet(string key, string value)
        {
            LocalDb.ExecNonQuery("INSERT OR REPLACE INTO Kv(K, V) VALUES(@k, @v);",
                LocalDb.P("@k", key), LocalDb.P("@v", value));
        }

        public static string? KvGet(string key)
        {
            return LocalDb.ExecScalar<string?>("SELECT V FROM Kv WHERE K = @k;", LocalDb.P("@k", key));
        }

        public static void IdentityEnsureEd25519(ref byte[]? publicKey, ref byte[]? secretKey)
        {
            var dt = LocalDb.Query("SELECT Ed25519Pub, Ed25519Priv FROM Identities WHERE Id = 1;");
            if (dt.Rows.Count > 0)
            {
                var row = dt.Rows[0];
                if (row["Ed25519Pub"] != DBNull.Value) publicKey = (byte[])row["Ed25519Pub"];
                if (row["Ed25519Priv"] != DBNull.Value) secretKey = (byte[])row["Ed25519Priv"];
            }

            if (publicKey == null || secretKey == null)
            {
                // Generate new Ed25519 key pair (placeholder - would use actual crypto library)
                publicKey = new byte[32]; // Placeholder
                secretKey = new byte[64]; // Placeholder

                LocalDb.ExecNonQuery("UPDATE Identities SET Ed25519Pub = @pub, Ed25519Priv = @priv WHERE Id = 1;",
                    LocalDb.P("@pub", publicKey), LocalDb.P("@priv", secretKey));
            }
        }

        public static byte[]? PeerGetEd25519(string peerName)
        {
            return LocalDb.ExecScalar<byte[]?>("SELECT Ed25519Pub FROM PeerKeys WHERE PeerName = @name AND IsRevoked = 0 ORDER BY Id DESC LIMIT 1;",
                LocalDb.P("@name", peerName));
        }

        public static void PeerSetEd25519_Tofu(string peerName, byte[] publicKey)
        {
            var now = DateTime.UtcNow.ToString("yyyy-MM-dd HH:mm:ss");
            LocalDb.ExecNonQuery(@"INSERT INTO PeerKeys(PeerName, Ed25519Pub, CreatedUtc, IsRevoked)
                                  VALUES(@name, @pub, @created, 0);",
                LocalDb.P("@name", peerName), LocalDb.P("@pub", publicKey), LocalDb.P("@created", now));
        }
    }

    /// <summary>
    /// Security-related database extensions
    /// </summary>
    public static class LocalDbExtensionsSecurity
    {
        public static void EnsurePeerExtraColumns()
        {
            // Schema migration would go here
        }

        public static DataTable PeerList()
        {
            return LocalDb.Query("SELECT Name, Trusted, CreatedUtc, LastSeenUtc FROM Peers ORDER BY Name;");
        }

        public static void PeerSetTrusted(string peerName, bool trusted)
        {
            LocalDb.ExecNonQuery("UPDATE Peers SET Trusted = @trusted WHERE Name = @name;",
                LocalDb.P("@name", peerName), LocalDb.P("@trusted", trusted ? 1 : 0));
        }

        public static void PeerSetNote(string peerName, string note)
        {
            LocalDb.ExecNonQuery("UPDATE Peers SET TrustNote = @note WHERE Name = @name;",
                LocalDb.P("@name", peerName), LocalDb.P("@note", note));
        }

        public static void PeerForgetEd25519(string peerName)
        {
            var now = DateTime.UtcNow.ToString("yyyy-MM-dd HH:mm:ss");
            LocalDb.ExecNonQuery("UPDATE PeerKeys SET IsRevoked = 1, RevokedUtc = @revoked WHERE PeerName = @name;",
                LocalDb.P("@name", peerName), LocalDb.P("@revoked", now));
        }

        public static void PeerMarkUnverified(string peerName)
        {
            LocalDb.ExecNonQuery("UPDATE Peers SET Verified = 0 WHERE Name = @name;",
                LocalDb.P("@name", peerName));
        }

        public static void PeerMarkVerified(string peerName)
        {
            LocalDb.ExecNonQuery("UPDATE Peers SET Verified = 1 WHERE Name = @name;",
                LocalDb.P("@name", peerName));
        }

        public static string? PeerGetField(string peerName, string fieldName)
        {
            var sql = $"SELECT {fieldName} FROM Peers WHERE Name = @name;";
            return LocalDb.ExecScalar<string?>(sql, LocalDb.P("@name", peerName));
        }

        public static void PeerSetPinned(string peerName, bool pinned)
        {
            // Placeholder for pinned status
        }

        public static bool PeerGetPinned(string peerName)
        {
            // Placeholder for pinned status
            return false;
        }

        public static DataTable GetSecurityEvents(string peerName)
        {
            return LocalDb.Query("SELECT EventType, Details, CreatedUtc FROM SecurityEvents WHERE PeerName = @name ORDER BY CreatedUtc DESC;",
                LocalDb.P("@name", peerName));
        }
    }
}